subprojects/workers/src/main/java/org/gradle/workers/internal/InProcessCompilerDaemonFactory.java

```diff
@@ -0,0 +1,144 @@
+                BuildOperationWorkerRegistry.Completion workerLease = parentWorkerOperation.operationStart();
+                BuildOperationDetails buildOperation = BuildOperationDetails.displayName(action.getDescription()).parent(parentBuildOperation).build();
+                try {
+                    return buildOperationExecutor.run(buildOperation, new Transformer<DefaultWorkResult, BuildOperationContext>() {
+                        @Override
+                        public DefaultWorkResult transform(BuildOperationContext buildOperationContext) {
+                            return executeInWorkerClassLoader(serverImplementationClass, action, spec, forkOptions.getClasspath(), forkOptions.getSharedPackages());
+                        }
+                    });
+                } finally {
+                    workerLease.operationFinish();
+                }
+            }
+        };
+    }
+
+    private <T extends WorkSpec> DefaultWorkResult executeInWorkerClassLoader(Class<? extends WorkerDaemonProtocol> serverImplementationClass, WorkerDaemonAction<T> action, T spec, Iterable<File> classpath, Iterable<String> sharedPackages) {
+        ClassLoader workerClassLoader = createWorkerClassLoader(serverImplementationClass, classpath, sharedPackages);
+        try {
+            LOGGER.info("Executing {} in in-process worker.", action.getDescription());
+            Callable<?> worker = transferWorkerIntoWorkerClassloader(action, spec, workerClassLoader);
+            Object result = worker.call();
+            DefaultWorkResult workResult = transferResultFromWorkerClassLoader(result);
+            LOGGER.info("Successfully executed {} in in-process worker.", action.getDescription());
+            return workResult;
+        } catch (Exception e) {
+            LOGGER.info("Exception executing {} in in-process worker: {}.", action.getDescription(), e);
+            throw UncheckedException.throwAsUncheckedException(e);
+        }
+    }
+
+    private ClassLoader createWorkerClassLoader(Class<? extends WorkerDaemonProtocol> serverImplementationClass, Iterable<File> classpath, Iterable<String> sharedPackages) {
+        ClassLoader isolatedWorkerClassLoader = classLoaderFactory.createIsolatedClassLoader(new DefaultClassPath(classpath));
+        FilteringClassLoader.Spec isolatedFilteringSpec = new FilteringClassLoader.Spec();
+        for (String sharedPackage : sharedPackages) {
+            isolatedFilteringSpec.allowPackage(sharedPackage);
+        }
+        ClassLoader filteredWorkerClassLoader = classLoaderFactory.createFilteringClassLoader(isolatedWorkerClassLoader, isolatedFilteringSpec);
+        ClassLoader workerServerClassLoader = serverImplementationClass.getClassLoader();
+        return new CachingClassLoader(new MultiParentClassLoader(filteredWorkerClassLoader, workerServerClassLoader));
```

Comment:
Thanks @adammurdoch, it took me a bit of time today to digest your feed back and do another round of understanding what's at stake. 

I reverted my changes because what was already in place was simply better and closer to what you describe than what I came up with. I renamed and moved things around though, so that things are easier to understand.

So, with this PR, compiler workers work just as before, and workers created through `WorkerExecutor` don't get any Gradle API out of the box. We'll improve that in another story.